File: 0357-count-numbers-with-unique-digits/0357-count-numbers-with-unique-digits.py
class Solution:
    def countNumbersWithUniqueDigits(self, n: int) -> int:
        
        
        if n == 0:
            return 1
        if n ==1 :
            return 10

        dp = [0] * (n+1)
        dp[0] = 1
        dp[1] = 10

        for i in range(2,n+1):
            tobeadded = 1

            k = i
            
            while k !=1:
                tobeadded *= (10 -(k-1))
                print(tobeadded)
                k-=1
            dp[i] = dp[i-1] + 9*tobeadded
            print(dp)
        return dp[-1]

File: 0357-count-numbers-with-unique-digits/0357-count-numbers-with-unique-digits.cpp
using namespace std;
#include <vector>
#include <iostream>
#include <iterator>



class Solution {
public:
    int countNumbersWithUniqueDigits(int n) {

        if (n == 0){
            return 1;
        }
        
        if (n == 1){
            return 10;
        }

        vector<int> dp(n+1,0);

        dp[0] = 1;
        dp[1] = 10;

        for (int i = 2; i < n+1; i ++){

            int tobeadded = 1;

            int k = i;
            while (k!= 1){
                tobeadded *= (10 - (k-1));
                k-=1;
            }
            dp[i] = dp[i-1] + 9 * tobeadded;

        }
        return dp.back();
        
    }
};


